pkg/cmd/server/origin/master.go

```diff
@@ -131,249 +134,210 @@ import (
 	}
 
-	if c.WebConsoleEnabled() {
-		handler = assetServerRedirect(handler, c.Options.AssetConfig.PublicURL)
-	}
+	if assetConfig != nil {
+		publicURL, err := url.Parse(assetConfig.Options.PublicURL)
+		if err != nil {
+			return nil, nil, err
+		}
+		messages = append(messages, fmt.Sprintf("Started Web Console %%s%s", publicURL.Path))
+	}
+
+	// TODO(sttts): resync with upstream handler chain and re-use upstream filters as much as possible
+	return func(apiHandler http.Handler, kc *genericapiserver.Config) (secure, insecure http.Handler) {
+		attributeGetter := kapiserverfilters.NewRequestAttributeGetter(c.RequestContextMapper)
+
+		handler := c.versionSkewFilter(apiHandler, c.getRequestContextMapper())
+		handler = c.authorizationFilter(handler)
+		handler = c.impersonationFilter(handler)
+
+		// audit handler must comes before the impersonationFilter to read the original user
+		if c.Options.AuditConfig.Enabled {
+			var writer io.Writer
+			if len(c.Options.AuditConfig.AuditFilePath) > 0 {
+				writer = &lumberjack.Logger{
+					Filename:   c.Options.AuditConfig.AuditFilePath,
+					MaxAge:     c.Options.AuditConfig.MaximumFileRetentionDays,
+					MaxBackups: c.Options.AuditConfig.MaximumRetainedFiles,
+					MaxSize:    c.Options.AuditConfig.MaximumFileSizeMegabytes,
+				}
+			} else {
+				// backwards compatible writer to regular log
+				writer = cmdutil.NewGLogWriterV(0)
+			}
+			handler = kapiserverfilters.WithAudit(handler, attributeGetter, writer)
+		}
+		handler = authenticationHandlerFilter(handler, c.Authenticator, c.getRequestContextMapper())
+		handler = namespacingFilter(handler, c.getRequestContextMapper())
+		handler = cacheControlFilter(handler, "no-store") // protected endpoints should not be cached
 
-	// Make the outermost filter the requestContextMapper to ensure all components share the same context
-	if contextHandler, err := kapi.NewRequestContextFilter(c.getRequestContextMapper(), handler); err != nil {
-		glog.Fatalf("Error setting up request context filter: %v", err)
-	} else {
-		handler = contextHandler
-	}
+		if c.Options.OAuthConfig != nil {
+			authConfig, err := BuildAuthConfig(c)
+			if err != nil {
+				glog.Fatalf("Failed to setup OAuth2: %v", err)
+			}
+			handler, err = authConfig.WithOAuth(handler)
+			if err != nil {
+				glog.Fatalf("Failed to setup OAuth2: %v", err)
+			}
+		}
 
-	longRunningRequestCheck := apiserver.BasicLongRunningRequestCheck(longRunningRE, map[string]string{"watch": "true"})
-	// TODO: MaxRequestsInFlight should be subdivided by intent, type of behavior, and speed of
-	// execution - updates vs reads, long reads vs short reads, fat reads vs skinny reads.
-	if c.Options.ServingInfo.MaxRequestsInFlight > 0 {
-		sem := make(chan bool, c.Options.ServingInfo.MaxRequestsInFlight)
-		handler = apiserver.MaxInFlightLimit(sem, longRunningRequestCheck, handler)
-	}
+		handler, err := assetConfig.WithAssets(handler)
```

Comment:
installing this as a handler filter is really odd to me... doesn't it mean all requests to any URL will have to fail against the asset mux before they bubble to '/' and are rehandled by the original handler?